src/core/utils/compile-template.ts

```diff
@@ -0,0 +1,120 @@
+): DocumentFragment {
+  const parser = new DOMParser();
+  const template: string = templatePugFn(props);
+  let elementBody: HTMLElement;
+
+  try {
+    elementBody = parser.parseFromString(template, 'text/html').body;
+
+    Object.keys(REGISTERED_COMPONENTS).forEach(componentName => {
+      const childElementTags = elementBody.querySelectorAll(componentName);
+
+      if (!childElementTags.length) {
+        return;
+      }
+
+      childElementTags.forEach((element: Element) => {
+        const dataName = element.getAttribute("data");
+
+        if (!dataName) {
+          console.error(`Attribute data was not specified in the markup for the ${element}`);
+          return;
+        }
+        const data = props[dataName] as Props;
+
+        if (data === undefined) {
+          console.error(`Property ${dataName} was not specified in the Props for the ${element}`);
+          return;
+        }
+
+        const path = getPathFromArray([pageEventName, dataName]);
+
+        if (isArray(data)) {
+          const childComponents = Object.values(data)
+            .map((value: Props) => {
+              const component = getComponent(componentName, pageEventName, dataName, value, events);
+
+              set(componentsState, path, component);
+
+              return component;
+            });
+
+          setAttributes(element, childComponents);
+        } else {
+          const childComponent = getComponent(componentName, pageEventName, dataName, data, events);
+
+          set(componentsState, path, childComponent);
+
+          setAttributes(element, [childComponent]);
+        }
+      })
+    });
+  } catch (err: unknown) {
+    throw new Error(`Template compilation failed due to ${String(err)}`);
+  }
+
+  const fragment = document.createDocumentFragment();
+
+  Array.from(elementBody.children).forEach(elem => {
+    fragment.appendChild(elem);
+  });
+
+  return fragment;
+}
+
+function getComponentInstance(componentName: string, props: unknown, eventName: string, events: Events): InstanceType<typeof Block> {
+  return new (REGISTERED_COMPONENTS as any)[componentName](props, eventName, events);
+}
+
+function setAttributes(childElementTag: Element, childComponents: InstanceType<typeof Block>[]): void {
+  const attributeNames = childElementTag.getAttributeNames();
+
+  const childElements = childComponents.map(childComponent => {
+    const childElement = childComponent.getContent();
+
+    attributeNames.forEach(attrName => {
+      const attrValue = childElementTag.getAttribute(attrName);
+
+      if (attrName === 'class') {
+        childElement.classList.add(attrValue ?? '');
+        return;
+      }
+
+      childElement.setAttribute(attrName, attrValue ?? '');
+    });
+
+    return childElement as Node;
+  });
+
+  childElementTag.replaceWith(...childElements);
+}
+
+function getComponent(componentName: string, pageEventName: string, dataName: string, props: Props, events: Events): InstanceType<typeof Block> {
```

Comment:
Такие длинные строки кода не очень удобно читать, можно отформатировать так:
```suggestion
function getComponent(
  componentName: string,
  pageEventName: string,
  dataName: string,
  props: Props,
  events: Events): InstanceType<typeof Block> {
```